compare_model_outputs_fx API implementation

DeepDriving opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49266 compare_model_outputs_fx API implementation**

compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

Commit:
compare_model_outputs_fx API implementation

compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+        output of linear layer in float model
+        """
+
+        def calibrate(model, calib_data):
```
I think you can use https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L81

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+        output of conv layer in float model
+        """
+
+        def calibrate(model, calib_data):
```
same here, you can use https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L81

DeepDriving replied:
Done

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+                for inp in calib_data:
+                    model(*inp)
+
+        def compare_and_validate_results(float_model, q_model, data):
```
the same function is defined in test_compare_model_outputs_conv_static_fx can you de-duplicate the code?

DeepDriving replied:
My preference will be keeping them separate for now (there is a slight difference between those two) and consider merging them down the road  when the api is more stable.

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+                for i, val in enumerate(v["quantized"]):
+                    self.assertTrue(v["float"][i].shape == v["quantized"][i].shape)
+
+            qengine = torch.backends.quantized.engine
```
is the indentation wrong here?

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+                for i, val in enumerate(v["quantized"]):
+                    self.assertTrue(v["float"][i].shape == v["quantized"][i].shape)
+
+            float_model = SingleLayerLinearModel()
```
same here, this is indented incorrectly I think, you need to remove one indent.

DeepDriving replied:
auto formatted the code.  Should took care of the indentation issue.

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
```
Could you also add the check for act_compare_dict.keys() as in test_numeric_suite.py?

DeepDriving replied:
Done

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
```
Could you also add the check for act_compare_dict.keys() as in test_numeric_suite.py?

DeepDriving replied:
Done

raghuramank100 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
         compare_and_validate_results(
             backup_prepared_model, q_model, module_swap_list, linear_data
         )
+
```
At a higher level can we merge the tests for numeric_suite and fx into a single file? The only difference in the tests would be that in one case, we quantize with fx APIs and call the fx numeric suite APIs. In the other case, we call with eager mode APIs.  This way, we can share the exact same models and reuse most of the testing code.

DeepDriving replied:
I would prefer keep them separate at this time, especially when we are still refining the APIs.  Fx and non-fx test cases use slightly different model definition. We can always merge them down the road.

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

DeepDriving commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            for k, v in act_compare_dict.items():
+                self.assertTrue(len(v["float"]) == len(v["quantized"]))
```
Done

DeepDriving commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            for k, v in act_compare_dict.items():
+                self.assertTrue(len(v["float"]) == len(v["quantized"]))
```
Done

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -237,3 +213,75 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+            expected_ob_dict_keys = {"stats"}
```
Expected key should have another one for fc1.

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -237,3 +213,75 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            expected_ob_dict_keys = {"stats"}
```
Expected key should have another one for conv.

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
+                input_node = node.args[0]
+                if (
+                    input_node.op == "call_function"
+                    and input_node.target.__name__ == "quantize_per_tensor"
```
you can do input_node.target == torch.quantize_per_tensor I think

DeepDriving replied:
Yes, it works.Thanks!

jerryzh168 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
+                    input_node.op == "call_function"
+                    and input_node.target.__name__ == "quantize_per_tensor"
+                ):
+                    target_dict[input_node.args[0].name + ".stats"] = modules[
```
can you add some comments, what is input_node.args[0]?

DeepDriving replied:
Done

jerryzh168 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
+                        node.target
+                    ].stats
+                else:
+                    target_dict[node.args[0].name + ".stats"] = modules[
```
please add a comment here as well describing what stats we are tracking here

DeepDriving replied:
Done

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -237,3 +213,74 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+        q_model = convert_fx(prepared_model)
+
+        linear_data = self.calib_data[0][0]
+        compare_and_validate_results(backup_prepared_model, q_model, linear_data)
```
May add one test to test the allow list?

jerryzh168 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -1,27 +1,22 @@
+    OutputLogger,
     ShadowLogger,
 )
+from torch.quantization.fx.quantization_patterns import NumericSuiteQuantizeHandler
```
can you define this handler in a different file?

DeepDriving replied:
Make sense. moving to numeric suite

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -762,3 +762,14 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
         quantizer.modules[node.target] = quantized_standalone_module
         return quantizer.quantized_graph.node_copy(node, load_arg(quantized=input_quantized_idxs))
+
+class NumericSuiteQuantizeHandler(QuantizeHandler):
```
can you define this in _numeric_suite_fx.py

DeepDriving replied:
Make sense as it is only needed for numeric suite.

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]

Commit:
Update on "compare_model_outputs_fx API implementation"


compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

[ghstack-poisoned]
